pkg/controller/utils/capability.go

```diff
@@ -0,0 +1,286 @@
+	if err != nil {
+		return &capability, fmt.Errorf("failed to get WorkloadDefinition %s: %v", def.Name, err)
+	}
+	def.TraitDefinition = traitDefinition
+	capability, err = util.ConvertTemplateJSON2Object(name, traitDefinition.Spec.Extension, traitDefinition.Spec.Template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert WorkloadDefinition to Capability Object")
+	}
+	return &capability, err
+}
+
+// GetOpenAPISchema gets OpenAPI v3 schema by TraitDefinition name
+func (def *CapabilityTraitDefinition) GetOpenAPISchema(ctx context.Context, k8sClient client.Client, namespace, name string) ([]byte, error) {
+	capability, err := def.GetCapabilityObject(ctx, k8sClient, namespace, name)
+	if err != nil {
+		return nil, err
+	}
+	return getOpenAPISchema(*capability)
+}
+
+// StoreOpenAPISchema stores OpenAPI v3 schema from TraitDefinition in ConfigMap
+func (def *CapabilityTraitDefinition) StoreOpenAPISchema(ctx context.Context, k8sClient client.Client, namespace, name string) error {
+	jsonSchema, err := def.GetOpenAPISchema(ctx, k8sClient, namespace, name)
+	if err != nil {
+		return fmt.Errorf(util.ErrGenerateOpenAPIV2JSONSchemaForCapability, def.Name, err)
+	}
+
+	traitDefinition := def.TraitDefinition
+	ownerReference := []metav1.OwnerReference{{
+		APIVersion:         traitDefinition.APIVersion,
+		Kind:               traitDefinition.Kind,
+		Name:               traitDefinition.Name,
+		UID:                traitDefinition.GetUID(),
+		Controller:         pointer.BoolPtr(true),
+		BlockOwnerDeletion: pointer.BoolPtr(true),
+	}}
+	return def.CreateOrUpdateConfigMap(ctx, k8sClient, traitDefinition.Name, jsonSchema, ownerReference)
+}
+
+// CapabilityBaseDefinition is the base struct for CapabilityWorkloadDefinition and CapabilityTraitDefinition
+type CapabilityBaseDefinition struct {
+}
+
+// CreateOrUpdateConfigMap creates ConfigMap to store OpenAPI v3 schema or or updates data in ConfigMap
+func (def *CapabilityBaseDefinition) CreateOrUpdateConfigMap(ctx context.Context, k8sClient client.Client, definitionName string, jsonSchema []byte, ownerReferences []metav1.OwnerReference) error {
+	cmName := fmt.Sprintf("%s%s", types.CapabilityConfigMapNamePrefix, definitionName)
+	ns := types.DefaultKubeVelaNS
+	var cm v1.ConfigMap
+	var data = map[string]string{
+		types.OpenapiV3JSONSchema: string(jsonSchema),
+	}
+	err := k8sClient.Get(ctx, client.ObjectKey{Namespace: ns, Name: cmName}, &cm)
+	if err != nil && apierrors.IsNotFound(err) {
+		cm = v1.ConfigMap{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "ConfigMap",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            cmName,
+				Namespace:       ns,
+				OwnerReferences: ownerReferences,
+			},
+			Data: data,
+		}
+		err = k8sClient.Create(ctx, &cm)
+		if err != nil {
+			return fmt.Errorf(util.ErrUpdateCapabilityInConfigMap, definitionName, err)
+		}
+		return nil
+	} else if cm.Data[types.OpenapiV3JSONSchema] != string(jsonSchema) {
```

Comment:
it already returned , so there's no need to write `else if`

Reply:
the jsonSchema will always diff in string format. you should not compare in this way

Reply:
How to compare string? I suppose I was comparing the string of josnSchema.

Reply:
you don't need to compare it, just update. If you really want to compare, you should decode it and compare structure.

Reply:
If there is diff, an update will happen, which will save time for the controller. Besides, there is not structure, the generated json schema is the type of `[]byte`. So it seems comparing the strings make sense.

Reply:
go map will in random order after decoded in string, so the string or byte compare not work here

Reply:
Okay, let's do update at any time.